prov/gni/src/gnix_mr.c

```diff
@@ -399,13 +415,252 @@ static int __gnix_destruct_registration(void *context)
+}
+
+static int __udreg_reg_mr(struct gnix_fid_domain *domain,
+		uint64_t                    address,
+		uint64_t                    length,
+		struct _gnix_fi_reg_context *fi_reg_context,
+		void                        **handle) {
+
+	return -FI_ENOSYS;
+}
+
+static int __udreg_dereg_mr(struct gnix_fid_domain *domain,
+		struct gnix_fid_mem_desc *md)
+{
+	return -FI_ENOSYS;
+}
+#endif
+
+struct gnix_mr_ops udreg_mr_ops = {
+	.init = __udreg_init,
+	.is_init = __udreg_is_init,
+	.reg_mr = __udreg_reg_mr,
+	.dereg_mr = __udreg_dereg_mr,
+};
+
+static int __cache_init(struct gnix_fid_domain *domain) {
+	return _gnix_mr_cache_init(&domain->mr_cache,
+			&domain->mr_cache_attr);
+}
+
+static int __cache_is_init(struct gnix_fid_domain *domain) {
+	return domain->mr_cache != NULL;
+}
+
+static int __cache_reg_mr(
+		struct gnix_fid_domain      *domain,
+		uint64_t                    address,
+		uint64_t                    length,
+		struct _gnix_fi_reg_context *fi_reg_context,
+		void                        **handle) {
+
+	return _gnix_mr_cache_register(domain->mr_cache, address, length,
+			fi_reg_context, handle);
+}
+
+static int __cache_dereg_mr(struct gnix_fid_domain *domain,
+		struct gnix_fid_mem_desc *md)
+{
+	return _gnix_mr_cache_deregister(domain->mr_cache, md);
+}
+
+struct gnix_mr_ops cache_mr_ops = {
+	.init = __cache_init,
+	.is_init = __cache_is_init,
+	.reg_mr = __cache_reg_mr,
+	.dereg_mr = __cache_dereg_mr,
+};
+
+int _gnix_open_cache(struct gnix_fid_domain *domain, int type)
+{
+	if (type < 0 || type >= GNIX_MR_MAX_TYPE)
+		return -FI_EINVAL;
+
+	if (domain->mr_ops)
+		return -FI_EBUSY;
+
+	switch(type) {
+	case GNIX_MR_TYPE_UDREG:
+		domain->mr_ops = &udreg_mr_ops;
+		break;
+	default:
+		domain->mr_ops = &cache_mr_ops;
+		break;
+	}
+
+	domain->mr_cache_type = type;
+	return 0;
+}
+
+int _gnix_close_cache(struct gnix_fid_domain *domain)
+{
+	int ret;
+
+	/* if the domain isn't being destructed by close, we need to check the
+	 * cache again. This isn't a likely case. Destroy must succeed since we
+	 * are in the destruct path */
+	if (domain->mr_cache) {
+		ret = _gnix_mr_cache_destroy(domain->mr_cache);
+		if (ret != FI_SUCCESS)
+			GNIX_FATAL(FI_LOG_DOMAIN, "failed to destroy mr cache "
+					"during domain destruct, dom=%p ret=%d\n",
+					domain, ret);
+	}
+
+#if HAVE_UDREG
+	if (domain->udreg_cache) {
+		ret = UDREG_CacheRelease(domain->udreg_cache);
+		if (unlikely(ret != UDREG_RC_SUCCESS))
+			GNIX_FATAL(FI_LOG_DOMAIN, "failed to release from "
+					"mr cache during domain destruct, dom=%p rc=%d\n",
+					domain, ret);
+
+		ret = UDREG_CacheDestroy(domain->udreg_cache);
+		if (unlikely(ret != UDREG_RC_SUCCESS))
+			GNIX_FATAL(FI_LOG_DOMAIN, "failed to destroy mr "
+					"cache during domain destruct, dom=%p rc=%d\n",
+					domain, ret);
+	}
+#endif
+
+	return 0;
+}
+
 gnix_mr_cache_attr_t _gnix_default_mr_cache_attr = {
 		.soft_reg_limit      = 4096,
 		.hard_reg_limit      = -1,
 		.hard_stale_limit    = 128,
```

Comment:
I think we talked about making these configurable via domain open_ops also.  Maybe open a separate issue for the domain open_ops stuff.